Form Designer Properties Pane Bug Fix

spikey-pb opened the pull request:
- Fix a bug where properties are erroneously applied to Explorer type gadgets on loading a form or switching between design and code view.

Commit:
Form Designer Properties Pane Bug Fix

- Fix a bug where properties are erroneously applied to Explorer type
  gadgets on loading a form or switching between design and code view.

fantaisie-software reviewed (commented):
Are you sure the 'ivalue' should match the real value flag ? We already have this value in the 'value' field. I think this 'ivalue' is needed because some real flags are composite of flags (especially on Windows)

spikey-pb commented:
I struggled to describe this comprehensibly, I guess I still haven't succeeded!  

I'm _not_ making 'iValue' the same as 'Value', I'm making 'iValues' for the same constant symbol name consistent across gadgets where these overlap.  So previously the 'iValue' for #PB_Explorer_NoFiles was 1 << 7 for ExplorerList and 1 << 4 for ExplorerTree.  This was leading to an extraneous matching problem when reparsing the code.  

Now the 'iValue' for #PB_Explorer_NoFiles will be consistently 1 << 13. 'Values' should be unaffected - unless I've made an error, but everything seemed to be working properly when I tested.  I will retest to make sure.

There may be a conflict of interest between this change and #321 though.  I need to retest #322 with #321 pulled to my repo. I also think these fields need new names!

fantaisie-software commented:
Ho I understand better now, thanks. It's indeed not obvious, that's why using raw number directly isn't a good idea. What about doing some 'EnumerationBinary' with replicated flags starting with #FI (FormInternal_) or something to replace these number and avoid future errors ?

```
EnumerationBinary
  #FI_Explorer_AlwaysShowSelection
  #FI_Explorer_AutoSort  
  #FI_Explorer_BorderLess
  #FI_Explorer_DrivesOnly
  #FI_Explorer_Editable  
  #FI_Explorer_FullRowSelect
  #FI_Explorer_GridLines    
  #FI_Explorer_HeaderDragDrop
  #FI_Explorer_MultiSelect   
  #FI_Explorer_NoButtons     
  #FI_Explorer_NoDirectoryChange
  #FI_Explorer_NoDriveRequester 
  #FI_Explorer_NoFiles          
  #FI_Explorer_NoFolders        
  #FI_Explorer_NoLines          
  #FI_Explorer_NoMyDocuments    
  #FI_Explorer_NoParentFolder   
  #FI_Explorer_NoSort               
EndEnumarationBinary
```

spikey-pb commented:
Sounds like a much better idea!  I'll have another go!  I'm thinking that the designer could benefit from some other code readability changes too.  I was thinking about posting a discussion article on the forum with a few suggestions to see what anyone else had thought of or bumped into.

spikey-pb committed:
Merge branch 'fantaisie-software:devel' into FDExplorerBug

spikey-pb commented:
What do you think about something that looks like:
```
Global NewList Gadgets.Gadgets()

; Add a new component to the list.
; TypeEnum: A #Form_Type_ constant from the list above. 
; IconEnum: An #IMAGE_FormIcons_ constant from the list in Common.pb.
Macro FDI_AddComponent(TypeEnum, IconEnum)
  AddElement(Gadgets()) : Gadgets()\type = TypeEnum : Gadgets()\icon = IconEnum
EndMacro

; Add a new flag to the current component.
; FlagName:    The literal string that should be displayed in the 'Properties' panel.
; LibSymbol:   The flag constant's symbol.
; PrivateEnum: The #FDI_ sequence constant.
Macro FDI_AddFlag(FlagName, LibSymbol, PrivateEnum)
  AddElement(Gadgets()\Flags()) 
  Gadgets()\Flags()\name = FlagName : Gadgets()\Flags()\value = LibSymbol : Gadgets()\Flags()\ivalue = PrivateEnum
EndMacro 

; Add a new event to the current component.
; EventName: The literal string that should be displayed in the 'Properties' panel.
Macro FDI_AddEvent(EventName)
  AddElement(Gadgets()\Events()) : Gadgets()\Events()\name = EventName
EndMacro 

;- Window
Enumeration
  #FDI_Window_SystemMenu
  #FDI_Window_MinimizeGadget
  #FDI_Window_MaximizeGadget
  #FDI_Window_SizeGadget
  #FDI_Window_Invisible
  #FDI_Window_TitleBar
  #FDI_Window_Tool
  #FDI_Window_BorderLess
  #FDI_Window_ScreenCentered
  #FDI_Window_WindowCentered
  #FDI_Window_Maximize
  #FDI_Window_Minimize
  #FDI_Window_NoGadgets
  #FDI_Window_NoActivate
EndEnumeration

FDI_AddComponent(#Form_Type_Window, #IMAGE_FormIcons_Calendar)
FDI_AddFlag("#PB_Window_SystemMenu", #PB_Window_SystemMenu, #FDI_Window_SystemMenu)
FDI_AddFlag("#PB_Window_MinimizeGadget", #PB_Window_MinimizeGadget, #FDI_Window_MinimizeGadget)
FDI_AddFlag("#PB_Window_MaximizeGadget", #PB_Window_MaximizeGadget, #FDI_Window_MaximizeGadget)
FDI_AddFlag("#PB_Window_SizeGadget", #PB_Window_SizeGadget, #FDI_Window_SizeGadget)
FDI_AddFlag("#PB_Window_Invisible", #PB_Window_Invisible, #FDI_Window_Invisible)
FDI_AddFlag("#PB_Window_TitleBar", #PB_Window_TitleBar, #FDI_Window_TitleBar)
FDI_AddFlag("#PB_Window_Tool", #PB_Window_Tool, #FDI_Window_Tool)
FDI_AddFlag("#PB_Window_BorderLess", #PB_Window_BorderLess, #FDI_Window_BorderLess)
FDI_AddFlag("#PB_Window_ScreenCentered", #PB_Window_ScreenCentered, #FDI_Window_ScreenCentered)
FDI_AddFlag("#PB_Window_WindowCentered", #PB_Window_WindowCentered, #FDI_Window_WindowCentered)
FDI_AddFlag("#PB_Window_Maximize", #PB_Window_Maximize, #FDI_Window_Maximize)
FDI_AddFlag("#PB_Window_Minimize", #PB_Window_Minimize, #FDI_Window_Minimize)
FDI_AddFlag("#PB_Window_NoGadgets", #PB_Window_NoGadgets, #FDI_Window_NoGadgets)
FDI_AddFlag("#PB_Window_NoActivate", #PB_Window_NoActivate, #FDI_Window_NoActivate)
FDI_AddEvent("#PB_Event_CloseWindow")
FDI_AddEvent("#PB_Event_Repaint")
FDI_AddEvent("#PB_Event_SizeWindow")
FDI_AddEvent("#PB_Event_MoveWindow")
FDI_AddEvent("#PB_Event_MinimizeWindow")
FDI_AddEvent("#PB_Event_MaximizeWindow")
FDI_AddEvent("#PB_Event_RestoreWindow")
FDI_AddEvent("#PB_Event_ActivateWindow")
FDI_AddEvent("#PB_Event_WindowDrop")
FDI_AddEvent("#PB_Event_Menu")
FDI_AddEvent("#PB_Event_Gadget")
FDI_AddEvent("#PB_Event_SysTray")
FDI_AddEvent("#PB_Event_Timer")
FDI_AddEvent("#PB_Event_GadgetDrop")

```

fantaisie-software commented:
Looks good to me. Just don't use a Macro if you don't need to as the generated code will be smaller and easier to debug.

spikey-pb commented:
I was hoping to improve the readability.  How do you feel about With…EndWith?

Commit:
Form Designer Properties Pane Bug Fix

- Fix a bug where properties are erroneously applied to Explorer type
  gadgets on loading a form or switching between design and code view.
- Use EnumerationBinary for better future proofing and readability.

fantaisie-software commented:
I mean you should use procedure instead. Gadget() is global IIRC, so you could just replace Macro with Procedure and we're good

fantaisie-software reviewed (approved):
Looks great, thanks !
